staging/src/k8s.io/cloud-provider/node/helpers/address.go

```diff
@@ -83,26 +84,58 @@ func PreferNodeIP(nodeIP net.IP, cloudNodeAddresses []v1.NodeAddress) ([]v1.Node
 		return sortedAddresses, nil
 	}
 
-	// For every address supplied by the cloud provider that matches nodeIP, nodeIP is the enforced node address for
-	// that address Type (like InternalIP and ExternalIP), meaning other addresses of the same Type are discarded.
-	// See #61921 for more information: some cloud providers may supply secondary IPs, so nodeIP serves as a way to
-	// ensure that the correct IPs show up on a Node object.
-	enforcedNodeAddresses := []v1.NodeAddress{}
+	// Otherwise the result is the same as for GetNodeAddressesFromNodeIP
+	return GetNodeAddressesFromNodeIP(nodeIP.String(), cloudNodeAddresses, false)
+}
+
+// GetNodeAddressesFromNodeIP filters the provided list of nodeAddresses to match the
+// providedNodeIP from the Node annotation (which is assumed to be non-empty). This is
+// used for external cloud providers.
+//
+// It will return node addresses filtered such that:
+//   - Any address matching nodeIP will be listed first.
+//   - If nodeIP matches an address of a particular type (internal or external),
+//     that will be the *only* address of that type returned.
+//   - All remaining addresses are listed after.
+//
+// (This does not have the same behavior with `0.0.0.0` and `::` as
+// GetNodeAddressesFromNodeIPLegacy, because that case never occurs for external cloud
+// providers, because kubelet does not set the `provided-node-ip` annotation in that
+// case.)
+func GetNodeAddressesFromNodeIP(providedNodeIP string, cloudNodeAddresses []v1.NodeAddress, allowDualStack bool) ([]v1.NodeAddress, error) {
+	nodeIPs, err := nodeutil.ParseNodeIPAnnotation(providedNodeIP, allowDualStack)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse node IP %q: %v", providedNodeIP, err)
+	}
 
+	enforcedNodeAddresses := []v1.NodeAddress{}
 	nodeIPTypes := make(map[v1.NodeAddressType]bool)
-	for _, nodeAddress := range cloudNodeAddresses {
-		if netutils.ParseIPSloppy(nodeAddress.Address).Equal(nodeIP) {
-			enforcedNodeAddresses = append(enforcedNodeAddresses, v1.NodeAddress{Type: nodeAddress.Type, Address: nodeAddress.Address})
-			nodeIPTypes[nodeAddress.Type] = true
+
+	for _, nodeIP := range nodeIPs {
+		// For every address supplied by the cloud provider that matches nodeIP,
+		// nodeIP is the enforced node address for that address Type (like
+		// InternalIP and ExternalIP), meaning other addresses of the same Type
+		// are discarded. See #61921 for more information: some cloud providers
+		// may supply secondary IPs, so nodeIP serves as a way to ensure that the
+		// correct IPs show up on a Node object.
+
+		matched := false
+		for _, nodeAddress := range cloudNodeAddresses {
+			if netutils.ParseIPSloppy(nodeAddress.Address).Equal(nodeIP) {
```

Comment:
At some point we need to verify which places actually need to be sloppy and which do not.  Sigh